File: src/codegopher/events/protocol.py
"""Typed JSONL protocol payloads for CodeGopher event streams."""

from __future__ import annotations

import json
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError

from codegopher.core.errors import CodeGopherError

PROTOCOL_VERSION = 1
REDACTED_VALUE = "[redacted]"
SENSITIVE_CONTAINER_KEYS = {"env", "headers", "headers_env"}
SENSITIVE_KEY_PARTS = (
    "api_key",
    "apikey",
    "authorization",
    "bearer",
    "credential",
    "password",
    "passwd",
    "secret",
    "token",
)


class ProtocolPayloadError(CodeGopherError):
    """Raised when a JSONL protocol payload cannot be decoded or validated."""


class ProtocolModel(BaseModel):
    """Base shape shared by all CodeGopher JSONL protocol messages."""

    model_config = ConfigDict(extra="forbid")

    version: Literal[1] = 1
    type: str = Field(min_length=1)
    session_id: str | None = None
    turn_id: str | None = None


class ProtocolCommand(ProtocolModel):
    """Base class for commands sent to the CodeGopher event subprocess."""


class ProtocolEvent(ProtocolModel):
    """Base class for events emitted by the CodeGopher event subprocess."""


class StartTurnCommand(ProtocolCommand):
    type: Literal["start_turn"] = "start_turn"
    prompt: str = Field(min_length=1)
    workspace_root: str = Field(min_length=1)
    selected_file: str | None = None
    editor_metadata: dict[str, Any] = Field(default_factory=dict)
    overrides: dict[str, Any] = Field(default_factory=dict)


class ApprovalResponseCommand(ProtocolCommand):
    type: Literal["approval_response"] = "approval_response"
    approval_id: str = Field(min_length=1)
    approved: bool
    reason: str | None = None


class CancelTurnCommand(ProtocolCommand):
    type: Literal["cancel_turn"] = "cancel_turn"
    turn_id: str = Field(min_length=1)


class ShutdownCommand(ProtocolCommand):
    type: Literal["shutdown"] = "shutdown"


class McpServerPayload(BaseModel):
    """Protocol representation of a configured MCP server."""

    model_config = ConfigDict(extra="forbid")

    enabled: bool = True
    transport: Literal["stdio", "sse"] = "stdio"
    command: str | None = None
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)
    cwd: str | None = None
    startup_timeout_seconds: float = Field(default=30.0, gt=0.0)
    url: str | None = None
    headers: dict[str, str] = Field(default_factory=dict)
    headers_env: dict[str, str] = Field(default_factory=dict)
    timeout_seconds: float = Field(default=5.0, gt=0.0)
    sse_read_timeout_seconds: float = Field(default=300.0, gt=0.0)


class GetEffectiveConfigCommand(ProtocolCommand):
    type: Literal["get_effective_config"] = "get_effective_config"
    workspace_root: str = Field(min_length=1)


class ListMcpServersCommand(ProtocolCommand):
    type: Literal["list_mcp_servers"] = "list_mcp_servers"
    workspace_root: str = Field(min_length=1)


class SaveMcpServerCommand(ProtocolCommand):
    type: Literal["save_mcp_server"] = "save_mcp_server"
    workspace_root: str = Field(min_length=1)
    server_name: str = Field(min_length=1)
    server: McpServerPayload


class SetMcpServerEnabledCommand(ProtocolCommand):
    type: Literal["set_mcp_server_enabled"] = "set_mcp_server_enabled"
    workspace_root: str = Field(min_length=1)
    server_name: str = Field(min_length=1)
    enabled: bool


class DeleteMcpServerCommand(ProtocolCommand):
    type: Literal["delete_mcp_server"] = "delete_mcp_server"
    workspace_root: str = Field(min_length=1)
    server_name: str = Field(min_length=1)


class SessionStartedEvent(ProtocolEvent):
    type: Literal["session_started"] = "session_started"
    session_id: str = Field(min_length=1)
    cwd: str = Field(min_length=1)
    provider: str = Field(min_length=1)
    model: str = Field(min_length=1)
    approval_mode: Literal["review", "auto", "yolo"]


class TurnStartedEvent(ProtocolEvent):
    type: Literal["turn_started"] = "turn_started"
    session_id: str = Field(min_length=1)
    turn_id: str = Field(min_length=1)
    cwd: str = Field(min_length=1)


class TextDeltaEvent(ProtocolEvent):
    type: Literal["text_delta"] = "text_delta"
    turn_id: str = Field(min_length=1)
    content: str = Field(min_length=1)


class ReasoningDeltaEvent(ProtocolEvent):
    type: Literal["reasoning_delta"] = "reasoning_delta"
    turn_id: str = Field(min_length=1)
    content: str = Field(min_length=1)


class ToolCallEvent(ProtocolEvent):
    type: Literal["tool_call"] = "tool_call"
    turn_id: str = Field(min_length=1)
    tool_id: str = Field(min_length=1)
    tool_name: str = Field(min_length=1)
    arguments_summary: str = ""


class ApprovalRequestEvent(ProtocolEvent):
    type: Literal["approval_request"] = "approval_request"
    turn_id: str = Field(min_length=1)
    approval_id: str = Field(min_length=1)
    tool_name: str = Field(min_length=1)
    arguments_summary: str = ""
    raw_arguments: dict[str, Any] | None = None


class ToolResultEvent(ProtocolEvent):
    type: Literal["tool_result"] = "tool_result"
    turn_id: str = Field(min_length=1)
    tool_id: str = Field(min_length=1)
    is_error: bool = False
    result_summary: str = ""


class ErrorEvent(ProtocolEvent):
    type: Literal["error"] = "error"
    code: str = Field(min_length=1)
    message: str = Field(min_length=1)


class TurnCompleteEvent(ProtocolEvent):
    type: Literal["turn_complete"] = "turn_complete"
    turn_id: str = Field(min_length=1)
    final_text: str = ""
    tool_count: int = Field(default=0, ge=0)
    approval_count: int = Field(default=0, ge=0)
    iteration_count: int = Field(default=0, ge=0)


class TaskContractStartedEvent(ProtocolEvent):
    type: Literal["task_contract_started"] = "task_contract_started"
    turn_id: str = Field(min_length=1)
    task_id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    status: Literal["active", "completed", "incomplete"]
    required_tool_calls: list[str] = Field(default_factory=list)
    required_artifacts: list[str] = Field(default_factory=list)


class TaskContractUpdatedEvent(ProtocolEvent):
    type: Literal["task_contract_updated"] = "task_contract_updated"
    turn_id: str = Field(min_length=1)
    task_id: str = Field(min_length=1)
    status: Literal["active", "completed", "incomplete"]
    observed_tool_calls: list[str] = Field(default_factory=list)
    observed_artifacts: list[str] = Field(default_factory=list)
    recovery_attempts: int = Field(default=0, ge=0)


class TaskContractGateFailedEvent(ProtocolEvent):
    type: Literal["task_contract_gate_failed"] = "task_contract_gate_failed"
    turn_id: str = Field(min_length=1)
    task_id: str = Field(min_length=1)
    gate_failures: list[str] = Field(default_factory=list)
    recovery_attempts: int = Field(default=0, ge=0)


class TaskContractCompletedEvent(ProtocolEvent):
    type: Literal["task_contract_completed"] = "task_contract_completed"
    turn_id: str = Field(min_length=1)
    task_id: str = Field(min_length=1)
    status: Literal["completed", "incomplete"]
    outcome: str | None = None


class ConfigSnapshotEvent(ProtocolEvent):
    type: Literal["config_snapshot"] = "config_snapshot"
    workspace_root: str = Field(min_length=1)
    provider: str = Field(min_length=1)
    model: str = Field(min_length=1)
    api_family: Literal["chat_completions", "responses"]
    base_url: str | None = None
    replay_reasoning_content: bool = False
    config_sources: list[str] = Field(default_factory=list)


class McpServerSnapshotPayload(BaseModel):
    """Redacted MCP server snapshot for VS Code display."""

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1)
    source: str | None = None
    server: McpServerPayload


class McpServersEvent(ProtocolEvent):
    type: Literal["mcp_servers"] = "mcp_servers"
    workspace_root: str = Field(min_length=1)
    servers: list[McpServerSnapshotPayload] = Field(default_factory=list)


class McpServerSavedEvent(ProtocolEvent):
    type: Literal["mcp_server_saved"] = "mcp_server_saved"
    workspace_root: str = Field(min_length=1)
    server_name: str = Field(min_length=1)
    server: McpServerPayload


class McpServerDeletedEvent(ProtocolEvent):
    type: Literal["mcp_server_deleted"] = "mcp_server_deleted"
    workspace_root: str = Field(min_length=1)
    server_name: str = Field(min_length=1)


_PROTOCOL_MODELS: dict[str, type[ProtocolModel]] = {
    "start_turn": StartTurnCommand,
    "approval_response": ApprovalResponseCommand,
    "cancel_turn": CancelTurnCommand,
    "shutdown": ShutdownCommand,
    "get_effective_config": GetEffectiveConfigCommand,
    "list_mcp_servers": ListMcpServersCommand,
    "save_mcp_server": SaveMcpServerCommand,
    "set_mcp_server_enabled": SetMcpServerEnabledCommand,
    "delete_mcp_server": DeleteMcpServerCommand,
    "session_started": SessionStartedEvent,
    "turn_started": TurnStartedEvent,
    "text_delta": TextDeltaEvent,
    "reasoning_delta": ReasoningDeltaEvent,
    "tool_call": ToolCallEvent,
    "approval_request": ApprovalRequestEvent,
    "tool_result": ToolResultEvent,
    "error": ErrorEvent,
    "turn_complete": TurnCompleteEvent,
    "task_contract_started": TaskContractStartedEvent,
    "task_contract_updated": TaskContractUpdatedEvent,
    "task_contract_gate_failed": TaskContractGateFailedEvent,
    "task_contract_completed": TaskContractCompletedEvent,
    "config_snapshot": ConfigSnapshotEvent,
    "mcp_servers": McpServersEvent,
    "mcp_server_saved": McpServerSavedEvent,
    "mcp_server_deleted": McpServerDeletedEvent,
}


def encode_jsonl_message(message: ProtocolModel) -> str:
    """Serialize one protocol model as a newline-terminated JSONL record."""

    return f"{message.model_dump_json()}\n"


def decode_jsonl_message(line: str) -> ProtocolModel:
    """Decode one JSONL protocol record into its typed payload model."""

    payload = line.strip()
    if not payload:
        raise ProtocolPayloadError("Empty protocol line")
    try:
        value = json.loads(payload)
    except json.JSONDecodeError as exc:
        raise ProtocolPayloadError(f"Malformed protocol JSON: {exc.msg}") from exc
    if not isinstance(value, dict):
        raise ProtocolPayloadError("Protocol payload must be a JSON object")

    version = value.get("version")
    if version is None:
        raise ProtocolPayloadError("Protocol payload missing version")
    if version != PROTOCOL_VERSION:
        raise ProtocolPayloadError(f"Unsupported protocol version: {version}")

    message_type = value.get("type")
    if message_type is None:
        raise ProtocolPayloadError("Protocol payload missing type")
    if not isinstance(message_type, str):
        raise ProtocolPayloadError("Protocol payload type must be a string")
    model_cls = _PROTOCOL_MODELS.get(message_type)
    if model_cls is None:
        raise ProtocolPayloadError(f"Unknown protocol type: {message_type}")

    try:
        return model_cls.model_validate(value)
    except ValidationError as exc:
        raise ProtocolPayloadError(f"Invalid {message_type} payload: {exc}") from exc


def redact_protocol_value(value: Any) -> Any:
    """Return a recursively redacted copy of a protocol trace value."""

    if isinstance(value, dict):
        redacted: dict[Any, Any] = {}
        for key, item in value.items():
            if not isinstance(key, str):
                redacted[key] = redact_protocol_value(item)
                continue
            normalized = key.lower().replace("-", "_")
            if normalized in SENSITIVE_CONTAINER_KEYS:
                redacted[key] = _redact_container_values(item)
            elif any(part in normalized for part in SENSITIVE_KEY_PARTS):
                redacted[key] = REDACTED_VALUE
            else:
                redacted[key] = redact_protocol_value(item)
        return redacted
    if isinstance(value, list):
        return [redact_protocol_value(item) for item in value]
    return value


def _redact_container_values(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: _redact_container_values(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_redact_container_values(item) for item in value]
    return REDACTED_VALUE

File: extensions/vscode/src/protocol.ts
export const protocolVersion = 1;
export const redactedProtocolValue = "[redacted]";

export type ProtocolVersion = typeof protocolVersion;
export type ApprovalMode = "review" | "auto" | "yolo";
export type ApiFamily = "chat_completions" | "responses";
export type McpTransport = "stdio" | "sse";

export interface ProtocolBase {
  version: ProtocolVersion;
  type: string;
  session_id?: string | null;
  turn_id?: string | null;
}

export interface StartTurnCommand extends ProtocolBase {
  type: "start_turn";
  prompt: string;
  workspace_root: string;
  selected_file?: string | null;
  editor_metadata?: Record<string, unknown>;
  overrides?: Record<string, unknown>;
}

export interface ApprovalResponseCommand extends ProtocolBase {
  type: "approval_response";
  approval_id: string;
  approved: boolean;
  reason?: string | null;
}

export interface CancelTurnCommand extends ProtocolBase {
  type: "cancel_turn";
  turn_id: string;
}

export interface ShutdownCommand extends ProtocolBase {
  type: "shutdown";
}

export interface GetEffectiveConfigCommand extends ProtocolBase {
  type: "get_effective_config";
  workspace_root: string;
}

export interface ListMcpServersCommand extends ProtocolBase {
  type: "list_mcp_servers";
  workspace_root: string;
}

export interface McpServerPayload {
  enabled?: boolean;
  transport?: McpTransport;
  command?: string | null;
  args?: string[];
  env?: Record<string, string>;
  cwd?: string | null;
  startup_timeout_seconds?: number;
  url?: string | null;
  headers?: Record<string, string>;
  headers_env?: Record<string, string>;
  timeout_seconds?: number;
  sse_read_timeout_seconds?: number;
}

export interface SaveMcpServerCommand extends ProtocolBase {
  type: "save_mcp_server";
  workspace_root: string;
  server_name: string;
  server: McpServerPayload;
}

export interface SetMcpServerEnabledCommand extends ProtocolBase {
  type: "set_mcp_server_enabled";
  workspace_root: string;
  server_name: string;
  enabled: boolean;
}

export interface DeleteMcpServerCommand extends ProtocolBase {
  type: "delete_mcp_server";
  workspace_root: string;
  server_name: string;
}

export type ProtocolCommand =
  | StartTurnCommand
  | ApprovalResponseCommand
  | CancelTurnCommand
  | ShutdownCommand
  | GetEffectiveConfigCommand
  | ListMcpServersCommand
  | SaveMcpServerCommand
  | SetMcpServerEnabledCommand
  | DeleteMcpServerCommand;

export interface SessionStartedEvent extends ProtocolBase {
  type: "session_started";
  session_id: string;
  cwd: string;
  provider: string;
  model: string;
  approval_mode: ApprovalMode;
}

export interface TurnStartedEvent extends ProtocolBase {
  type: "turn_started";
  session_id: string;
  turn_id: string;
  cwd: string;
}

export interface TextDeltaEvent extends ProtocolBase {
  type: "text_delta";
  turn_id: string;
  content: string;
}

export interface ReasoningDeltaEvent extends ProtocolBase {
  type: "reasoning_delta";
  turn_id: string;
  content: string;
}

export interface ToolCallEvent extends ProtocolBase {
  type: "tool_call";
  turn_id: string;
  tool_id: string;
  tool_name: string;
  arguments_summary?: string;
}

export interface ApprovalRequestEvent extends ProtocolBase {
  type: "approval_request";
  turn_id: string;
  approval_id: string;
  tool_name: string;
  arguments_summary?: string;
  raw_arguments?: Record<string, unknown> | null;
}

export interface ToolResultEvent extends ProtocolBase {
  type: "tool_result";
  turn_id: string;
  tool_id: string;
  is_error?: boolean;
  result_summary?: string;
}

export interface ErrorEvent extends ProtocolBase {
  type: "error";
  code: string;
  message: string;
}

export interface TurnCompleteEvent extends ProtocolBase {
  type: "turn_complete";
  turn_id: string;
  final_text?: string;
  tool_count?: number;
  approval_count?: number;
  iteration_count?: number;
}

export interface ConfigSnapshotEvent extends ProtocolBase {
  type: "config_snapshot";
  workspace_root: string;
  provider: string;
  model: string;
  api_family: ApiFamily;
  base_url?: string | null;
  replay_reasoning_content?: boolean;
  config_sources?: string[];
}

export interface McpServerSnapshotPayload {
  name: string;
  source?: string | null;
  server: McpServerPayload;
}

export interface McpServersEvent extends ProtocolBase {
  type: "mcp_servers";
  workspace_root: string;
  servers?: McpServerSnapshotPayload[];
}

export interface McpServerSavedEvent extends ProtocolBase {
  type: "mcp_server_saved";
  workspace_root: string;
  server_name: string;
  server: McpServerPayload;
}

export interface McpServerDeletedEvent extends ProtocolBase {
  type: "mcp_server_deleted";
  workspace_root: string;
  server_name: string;
}

export type ProtocolEvent =
  | SessionStartedEvent
  | TurnStartedEvent
  | TextDeltaEvent
  | ReasoningDeltaEvent
  | ToolCallEvent
  | ApprovalRequestEvent
  | ToolResultEvent
  | ErrorEvent
  | TurnCompleteEvent
  | ConfigSnapshotEvent
  | McpServersEvent
  | McpServerSavedEvent
  | McpServerDeletedEvent;

export type ProtocolMessage = ProtocolCommand | ProtocolEvent;

export const protocolTypes = [
  "start_turn",
  "approval_response",
  "cancel_turn",
  "shutdown",
  "get_effective_config",
  "list_mcp_servers",
  "save_mcp_server",
  "set_mcp_server_enabled",
  "delete_mcp_server",
  "session_started",
  "turn_started",
  "text_delta",
  "reasoning_delta",
  "tool_call",
  "approval_request",
  "tool_result",
  "error",
  "turn_complete",
  "config_snapshot",
  "mcp_servers",
  "mcp_server_saved",
  "mcp_server_deleted"
] as const;

type ProtocolType = (typeof protocolTypes)[number];

export const protocolEventTypes = [
  "session_started",
  "turn_started",
  "text_delta",
  "reasoning_delta",
  "tool_call",
  "approval_request",
  "tool_result",
  "error",
  "turn_complete",
  "config_snapshot",
  "mcp_servers",
  "mcp_server_saved",
  "mcp_server_deleted"
] as const;

const protocolTypeSet = new Set<string>(protocolTypes);
const protocolEventTypeSet = new Set<string>(protocolEventTypes);
const sensitiveContainerKeys = new Set(["env", "headers", "headers_env"]);
const sensitiveKeyParts = [
  "api_key",
  "apikey",
  "authorization",
  "bearer",
  "credential",
  "password",
  "passwd",
  "secret",
  "token"
] as const;

export class ProtocolParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolParseError";
  }
}

export function encodeProtocolMessage(message: ProtocolMessage): string {
  return `${JSON.stringify({ ...message, version: protocolVersion })}\n`;
}

export function decodeProtocolLine(line: string): ProtocolMessage {
  const payload = line.trim();
  if (!payload) {
    throw new ProtocolParseError("Empty protocol line");
  }

  let value: unknown;
  try {
    value = JSON.parse(payload);
  } catch (error) {
    const detail = error instanceof Error ? error.message : "invalid JSON";
    throw new ProtocolParseError(`Malformed protocol JSON: ${detail}`);
  }

  const message = asRecord(value, "Protocol payload must be a JSON object");
  if (message.version !== protocolVersion) {
    if (!Object.hasOwn(message, "version")) {
      throw new ProtocolParseError("Protocol payload missing version");
    }
    throw new ProtocolParseError(`Unsupported protocol version: ${String(message.version)}`);
  }

  const type = message.type;
  if (typeof type !== "string") {
    if (!Object.hasOwn(message, "type")) {
      throw new ProtocolParseError("Protocol payload missing type");
    }
    throw new ProtocolParseError("Protocol payload type must be a string");
  }
  if (!protocolTypeSet.has(type)) {
    throw new ProtocolParseError(`Unknown protocol type: ${type}`);
  }

  validateMessage(type as ProtocolType, message);
  return message as unknown as ProtocolMessage;
}

export class JsonlProtocolParser {
  private buffer = "";

  push(chunk: Buffer | string): ProtocolMessage[] {
    this.buffer += typeof chunk === "string" ? chunk : chunk.toString("utf8");
    const messages: ProtocolMessage[] = [];

    for (;;) {
      const newlineIndex = this.buffer.indexOf("\n");
      if (newlineIndex === -1) {
        break;
      }
      const line = stripTrailingCarriageReturn(this.buffer.slice(0, newlineIndex));
      this.buffer = this.buffer.slice(newlineIndex + 1);
      messages.push(decodeProtocolLine(line));
    }

    return messages;
  }

  flush(): ProtocolMessage[] {
    if (!this.buffer) {
      return [];
    }
    const line = stripTrailingCarriageReturn(this.buffer);
    this.buffer = "";
    return [decodeProtocolLine(line)];
  }

  clear(): void {
    this.buffer = "";
  }
}

export function isProtocolEvent(message: ProtocolMessage): message is ProtocolEvent {
  return protocolEventTypeSet.has(message.type);
}

export function redactProtocolValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => redactProtocolValue(item));
  }
  if (!isRecord(value)) {
    return value;
  }

  const redacted: Record<string, unknown> = {};
  for (const [key, item] of Object.entries(value)) {
    const normalized = key.toLowerCase().replaceAll("-", "_");
    if (sensitiveContainerKeys.has(normalized)) {
      redacted[key] = redactContainerValues(item);
    } else if (sensitiveKeyParts.some((part) => normalized.includes(part))) {
      redacted[key] = redactedProtocolValue;
    } else {
      redacted[key] = redactProtocolValue(item);
    }
  }
  return redacted;
}

function redactContainerValues(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => redactContainerValues(item));
  }
  if (!isRecord(value)) {
    return redactedProtocolValue;
  }
  return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, redactContainerValues(item)]));
}

function validateMessage(type: ProtocolType, message: Record<string, unknown>): void {
  switch (type) {
    case "start_turn":
      requireString(message, "prompt");
      requireString(message, "workspace_root");
      optionalRecord(message, "editor_metadata");
      optionalRecord(message, "overrides");
      optionalStringOrNull(message, "selected_file");
      break;
    case "approval_response":
      requireString(message, "approval_id");
      requireBoolean(message, "approved");
      optionalStringOrNull(message, "reason");
      break;
    case "cancel_turn":
      requireString(message, "turn_id");
      break;
    case "shutdown":
      break;
    case "get_effective_config":
    case "list_mcp_servers":
      requireString(message, "workspace_root");
      break;
    case "save_mcp_server":
      requireString(message, "workspace_root");
      requireString(message, "server_name");
      validateMcpServer(asRecord(message.server, "server must be an object"));
      break;
    case "set_mcp_server_enabled":
      requireString(message, "workspace_root");
      requireString(message, "server_name");
      requireBoolean(message, "enabled");
      break;
    case "delete_mcp_server":
      requireString(message, "workspace_root");
      requireString(message, "server_name");
      break;
    case "session_started":
      requireString(message, "session_id");
      requireString(message, "cwd");
      requireString(message, "provider");
      requireString(message, "model");
      requireOneOf(message, "approval_mode", ["review", "auto", "yolo"]);
      break;
    case "turn_started":
      requireString(message, "session_id");
      requireString(message, "turn_id");
      requireString(message, "cwd");
      break;
    case "text_delta":
    case "reasoning_delta":
      requireString(message, "turn_id");
      requireString(message, "content");
      break;
    case "tool_call":
      requireString(message, "turn_id");
      requireString(message, "tool_id");
      requireString(message, "tool_name");
      optionalString(message, "arguments_summary");
      break;
    case "approval_request":
      requireString(message, "turn_id");
      requireString(message, "approval_id");
      requireString(message, "tool_name");
      optionalString(message, "arguments_summary");
      optionalRecordOrNull(message, "raw_arguments");
      break;
    case "tool_result":
      requireString(message, "turn_id");
      requireString(message, "tool_id");
      optionalBoolean(message, "is_error");
      optionalString(message, "result_summary");
      break;
    case "error":
      requireString(message, "code");
      requireString(message, "message");
      break;
    case "turn_complete":
      requireString(message, "turn_id");
      optionalString(message, "final_text");
      optionalNonNegativeNumber(message, "tool_count");
      optionalNonNegativeNumber(message, "approval_count");
      optionalNonNegativeNumber(message, "iteration_count");
      break;
    case "config_snapshot":
      requireString(message, "workspace_root");
      requireString(message, "provider");
      requireString(message, "model");
      requireOneOf(message, "api_family", ["chat_completions", "responses"]);
      optionalStringOrNull(message, "base_url");
      optionalBoolean(message, "replay_reasoning_content");
      optionalStringArray(message, "config_sources");
      break;
    case "mcp_servers":
      requireString(message, "workspace_root");
      optionalMcpServerSnapshots(message, "servers");
      break;
    case "mcp_server_saved":
      requireString(message, "workspace_root");
      requireString(message, "server_name");
      validateMcpServer(asRecord(message.server, "server must be an object"));
      break;
    case "mcp_server_deleted":
      requireString(message, "workspace_root");
      requireString(message, "server_name");
      break;
  }
}

function validateMcpServer(server: Record<string, unknown>): void {
  optionalBoolean(server, "enabled");
  optionalOneOf(server, "transport", ["stdio", "sse"]);
  optionalStringOrNull(server, "command");
  optionalStringArray(server, "args");
  optionalStringMap(server, "env");
  optionalStringOrNull(server, "cwd");
  optionalPositiveNumber(server, "startup_timeout_seconds");
  optionalStringOrNull(server, "url");
  optionalStringMap(server, "headers");
  optionalStringMap(server, "headers_env");
  optionalPositiveNumber(server, "timeout_seconds");
  optionalPositiveNumber(server, "sse_read_timeout_seconds");
}

function optionalMcpServerSnapshots(message: Record<string, unknown>, key: string): void {
  if (!Object.hasOwn(message, key)) {
    return;
  }
  const value = message[key];
  if (!Array.isArray(value)) {
    throw new ProtocolParseError(`${key} must be an array`);
  }
  for (const item of value) {
    const snapshot = asRecord(item, `${key} entries must be objects`);
    requireString(snapshot, "name");
    optionalStringOrNull(snapshot, "source");
    validateMcpServer(asRecord(snapshot.server, "server must be an object"));
  }
}

function asRecord(value: unknown, message: string): Record<string, unknown> {
  if (!isRecord(value)) {
    throw new ProtocolParseError(message);
  }
  return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function requireString(message: Record<string, unknown>, key: string): void {
  const value = message[key];
  if (typeof value !== "string" || value.length === 0) {
    throw new ProtocolParseError(`${key} must be a non-empty string`);
  }
}

function requireBoolean(message: Record<string, unknown>, key: string): void {
  if (typeof message[key] !== "boolean") {
    throw new ProtocolParseError(`${key} must be a boolean`);
  }
}

function requireOneOf(message: Record<string, unknown>, key: string, choices: readonly string[]): void {
  const value = message[key];
  if (typeof value !== "string" || !choices.includes(value)) {
    throw new ProtocolParseError(`${key} must be one of ${choices.join(", ")}`);
  }
}

function optionalString(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key) && typeof message[key] !== "string") {
    throw new ProtocolParseError(`${key} must be a string`);
  }
}

function optionalStringOrNull(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key) && message[key] !== null && typeof message[key] !== "string") {
    throw new ProtocolParseError(`${key} must be a string or null`);
  }
}

function optionalBoolean(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key) && typeof message[key] !== "boolean") {
    throw new ProtocolParseError(`${key} must be a boolean`);
  }
}

function optionalRecord(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key)) {
    asRecord(message[key], `${key} must be an object`);
  }
}

function optionalRecordOrNull(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key) && message[key] !== null) {
    asRecord(message[key], `${key} must be an object or null`);
  }
}

function optionalStringArray(message: Record<string, unknown>, key: string): void {
  if (!Object.hasOwn(message, key)) {
    return;
  }
  const value = message[key];
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string")) {
    throw new ProtocolParseError(`${key} must be an array of strings`);
  }
}

function optionalStringMap(message: Record<string, unknown>, key: string): void {
  if (!Object.hasOwn(message, key)) {
    return;
  }
  const value = asRecord(message[key], `${key} must be an object`);
  for (const [mapKey, mapValue] of Object.entries(value)) {
    if (typeof mapKey !== "string" || typeof mapValue !== "string") {
      throw new ProtocolParseError(`${key} must be an object of strings`);
    }
  }
}

function optionalOneOf(message: Record<string, unknown>, key: string, choices: readonly string[]): void {
  if (!Object.hasOwn(message, key)) {
    return;
  }
  requireOneOf(message, key, choices);
}

function optionalPositiveNumber(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key) && !isPositiveNumber(message[key])) {
    throw new ProtocolParseError(`${key} must be a positive number`);
  }
}

function optionalNonNegativeNumber(message: Record<string, unknown>, key: string): void {
  if (Object.hasOwn(message, key) && !isNonNegativeNumber(message[key])) {
    throw new ProtocolParseError(`${key} must be a non-negative number`);
  }
}

function isPositiveNumber(value: unknown): boolean {
  return typeof value === "number" && Number.isFinite(value) && value > 0;
}

function isNonNegativeNumber(value: unknown): boolean {
  return typeof value === "number" && Number.isFinite(value) && value >= 0;
}

function stripTrailingCarriageReturn(value: string): string {
  return value.endsWith("\r") ? value.slice(0, -1) : value;
}
